File: javascript/03_chapter/SequentialSearchST.js
/**
 * 顺序查找（基于无序列表）
 */

class Node {
    constructor(key, val, next){
        this.key = key;
        this.value = val;
        this.next = next;
    }
}

class SequentialSearchST {
    constructor(){
        this.first = null;
    }

    get(key){
        for (let x = this.first; x != null; x = x.next){
            if (x.key === key){
                return x.value
            }
        }
        return null;
    }

    put(key, val){
        for (let x = this.first; x != null; x = x.next){
            if (x.key === key){
                x.value = val;
                return;
            }
        }
        this.first = new Node(key, val, this.first);
    }
}

let dict = {
    'S': 0,
    'E': 1,
    'A': 2,
    'R': 3,
    'C': 4,
    'H': 5,
    'E': 6,
    'X': 7,
    'A': 8,
    'M': 9,
    'P': 10,
    'L': 11,
    'E': 12
};

let st = new SequentialSearchST();
let keys = Object.keys(dict);
for (let key of keys){
    st.put(key, dict[key]);
}

for (let key of keys){
    console.log(key, st.get(key))
}

File: algs4/03_chapter/SequentialSearchST.java
/**
 * 3.1.4 顺序查找（基于无序列表）
 * 编译： javac -cp :../stdlib.jar:../01_chapter SequentialSearchST.java
 * 运行： java -cp :../stdlib.jar:../01_chapter SequentialSearchST.java < tinyST.txt
 * 结果：
 * L 11
 * P 10
 * M 9
 * X 7
 * H 5
 * C 4
 * R 3
 * A 8
 * E 12
 * S 0
 */

public class SequentialSearchST<Key, Value>{
    private Node first;

    private class Node {
        Key key;
        Value val;
        Node next;
        public Node(Key key, Value val, Node next){
            this.key = key;
            this.val = val;
            this.next = next;
        }
    }

    public Value get(Key key) {
        for (Node x = first; x != null; x = x.next){
            if (key.equals(x.key)){
                return x.val;
            }
        }
        return null;
    }

    public void put(Key key, Value val){
        for (Node x = first; x != null; x = x.next){
            if (key.equals(x.key)){
                x.val = val;
                return;
            }
        }
        first = new Node(key, val, first);
    }

    public Iterable<Key> keys(){
        Queue<Key> queue = new Queue<Key>();
        for (Node x = first; x != null; x = x.next){
            queue.enqueue(x.key);
        }
        return queue;
    }

    public static void main(String[] args){
        SequentialSearchST<String, Integer> st = new SequentialSearchST<String, Integer>();
        for (int i = 0; !StdIn.isEmpty(); i++){
            String key = StdIn.readString();
            st.put(key, i);
        }
        for (String s : st.keys()){
            StdOut.println(s + " " + st.get(s));
        }
    }
}